### ISSUE TITLE
Please allow defining types separate from implementations

### ISSUE THREAD

**Issue by Gozala** [facebook/flow#838]
I constantly find myself wishing I could define types separately from their implementations kind of like you do it in haskell. Unfortunately now type declarations for same named functions are completely dropped see #748 

I would like to illustrate it in the example:

```js
import {html, forward, Effects} from "reflex";
import type {Task, Update, View} from "reflex";

type Model = {value: number};

type init = () => Model;
export var init = () => {value: 0};

type Increment = {type: 'Increment'};
type Decrement = {type: 'Decrement'};
type Action 
  = Increment
  | Decrement;


type inc = () => Increment;
export var inc = () => {type: 'Increment'};

type dec = () => Decrement;
export var dec = () => {type: 'Decrement'};

type update = Update<Model, Action>;
// Imported Update has following type signature:
// type Update <Model, Action> = (Model, Action) => [Model, Task<Never, Action>]

export var update = (model, action) =>
  action instanceof Increment ?
    [set('x', model.x + 1, model), Effects.none] :
  action instanceof Decrement ?
    [set('x', model.x - 1, model), Effects.none] :
    [model, Effects.none];

// View
type view = View<[Model, Address<Action>]>;

// Where View has following type:
// type View <Args> = (...Args) => VirtualElement;
export var view = (model, address) => {
  return html.span({key: 'counter'}, [
    html.button({
      key: 'decrement',
      onClick: forward(Decrement, address)
    }, ["-"]),
    html.span({
      key: 'value'
    }, [String(model.value)]),
    html.button({
      key: 'increment',
      onClick: forward(Increment, address)
    }, ["+"])
  ]);
};
```

Interesting parts are `Update` and `View` imported type, that were used to create type definitions for `update` and `view` functions. I find it a lot more handy than and clear than annotating individual functions. I also find it a lot more convenient to be able to annotate things without changing them in place. Note that I also have dropped mandatory argument names in the function type definitions which I find redundant when defining types separately.

**Comment by samwgoldman** [comment]
Can you please make a minimal example to show what you want? It would really help if you also simplified the problem domain from something complex (the reflex API) to something trivial.

**Comment by Gozala** [comment]
> Can you please make a minimal example to show what you want? It would really help if you also simplified the problem domain from something complex (the reflex API) to something trivial.

Sorry I should not have just dumped that large example and expect that it would be obvious, let me try to walk through it and pull out small sections that could serve as simpler examples. Although please keep in mind that I do find this feature useful especially when dealing with more complex APIs, I'll try to elaborate on this below.

Here is a very basic example of how type signatures and functions could be defined separately:

```js
type Model = {value: number};

type init = () => Model;
export var init = () => {value: 0};
```

While I do think it's a lot nicer to have a type definitions separate from function definition in the above example it does not really buys much.

Now consider following example where function needs to implement more complex type signature:

```js
type Effect <Action> = {$typeof: "Effect", perform: () => Action}
type Update <Model, Action> = (Model, Action) => [Model, Effect<Action>]

type Counter = {value: number}
type Action 
  = {$typeof: "Increment"}
  | {$typeof: "Decrement"}

type update = Update<Counter, Action>
const update = (model, action) =>
  action.$typeof === "Increment" ? [{value: model.value + 1},  Effects.none]:
  action.$typeof === "Decrement" ? {value: model.value - 1} , Effects.none]:
  [model, Effects.none]
```

Please try not to get hung up on it too much on the specifics of the API or the `Effect` type my primary intent here is to provide an example where polymorphic type of the function incorporates other polymorphic types. The important piece here is that libraries / frameworks often times require users to implement functions with a specific signature in my case it is `Update` type signature and I would very much like if user could import that polymorphic `Update` type and claim that it's provided functions comply to it by parameterizing instead of inlining complex API signatures in every function.

Sure one could define type for required functions return type to make things little easier, bug that would still require inlining argument types etc..

Does this examples with more explanations make more sense ?

P.S.: I dropped argument name in function type signatures as I think they are redundant when types are defined separately.

**Comment by Gozala** [comment]
Another reason why I would like to have a way to define types separate form implementations is because for the project I'm working I can't just use flow as is my only option is to use [flow through comments](http://flowtype.org/blog/2015/02/20/Flow-Comments.html) and having types separate from implementation really makes code more readable & compact.

**Comment by Gozala** [comment]
Another great side effect of this would that one will be able to define types in a separate file from the implementation. Although figuring out how to associating them will need to be determined, but even without any changes it would greatly improve "through comment" use case as required comments could be reduced to just:

```js
// import type {init, update} from "./counter.type.js"

export const init = () => {value: 0};
export const update = (model, action) =>
  action.$typeof === "Increment" ? [{value: model.value + 1},  Effects.none]:
  action.$typeof === "Decrement" ? {value: model.value - 1} , Effects.none]:
  [model, Effects.none]
```

while `./counter.type.js` could have all the type definitions like:

```js
type Counter = {value: number};

export type init = () => Model;

type Effect <Action> = {$typeof: "Effect", perform: () => Action}
type Update <Model, Action> = (Model, Action) => [Model, Effect<Action>]

type Counter = {value: number}
type Action 
  = {$typeof: "Increment"}
  | {$typeof: "Decrement"}

export type update = Update<Counter, Action>
```

**Comment by Gozala** [comment]
I actually thought of a much better example to illustrate my point:

```js
type Interpolation<a> = (a, a, Float) => a

type Float = number
type float = Interpolation<Float>
const float = (from, to, progress) =>
  from + (to - from) * progress

type Point2D = {$typeof:"Point2D", x:Float, y:Float}
type point2D = Interpolation<Point2D>
const point2D = (from, to, progress) => ({
  $typeof:"Point2D"
  x: float(from.x, to.x, progress),
  y: float(from.y, to.y, progress)
})

// ...


type point2D = Interpolation<Point3D>

// ...

type color = Interpolation<Color>

// ...

type Easing = Float => Float
type ease <a> = (Easing, Interpolation<a>, a, a, Time, Time) => a
const ease = (easing, interpolation, from, to, duration, time) =>
  interpolation(from, to, easing(Math.min(time / duration, 1)))
```

1. It's a lot more convenient to define polymorphic `Interpolation` type and then annotate functions as parameterized `Interpolation` instead of repeating same type signature over and over.

2. In order to define polymorphic `ease` function you need a polymorphic `Interpolation` type definition anyhow in which case it makes a lot more sense to use that `Interpolation` type annotation for variants of `interpolation` instead of repeating a signature and expect that users will notice it's the same.

**Comment by Gozala** [comment]
It seems that following code actually type checks fine:

```js
/* @flow */


type Float = number
type Interpolation<a> = (from:a, to:a, progress:Float) => a

export const float:Interpolation<Float> = (from, to, progress) =>
  from + (to - from) * progress
```

**Comment by Gozala** [comment]
I turns out pattern mentioned above can be used as a poor mans way to separate type definitions from implementations. Here is an example of this:

## ./type/easing.js
```js
/* @flow */

export type Float = number
export type Interpolation<a> = (from:a, to:a, progress:Float) => a;
export type float = Interpolation<Float>;
```

## ./src/easing.js
```js
/* @flow */

/*:: import * as type from "../type/flowing" */

export const float/*:type.float*/ = (from, to, progress) =>
  from + (to - from)
```

**Comment by gyzerok** [comment]
I absolute like this approach too. Seems following looks much cleaner than original syntax.
```js
type sum = (Number, Number) => Number
const sum = (a, b) => a + b;
// or
export type sum = (Number, Number) => Number
export const sum = (a, b) => a + b;
```

**Comment by Gozala** [comment]
@gyzerok I don't think it should be one or the other current syntax and new could happily co-exist, in fact current syntax is helpful for inlined callbacks ;)

To summarize current state of this issue (as I understand it). Some constraint  were expressed in regards to using type aliases for type annotating functions:

> it comes down to a separate understanding of what the `type foo = ...` syntax *means* when I see that line, I see it as "a universe of values is described by this type" and your recommendation seems to imply a very different interpretation, which is "this type describes a single type"

@samwgoldman also generously offered a help to get oriented around the code if someone wants to take a stab at it.

Unfortunately I have never programmed in ocaml & I still have several blockers preventing from integrating flow into my employers toolchain. This makes it very difficult to justify hacking on this. I'll let this thread know things change & I'll have a chance to work on this.

**Comment by gyzerok** [comment]
@Gozala Unfortunely I do not have experience with OCaml too.
I do not understand your thoughts on inline callbacks since you define their types not on the caller side.

```js
// When you define
type map<U, V> = (U => V, Array<U>) => Array<V>
const map = (predicate, source) => ...

// When you use
const squares = map(x => x * x, [1, 2, 3]);
```

**Comment by jasonkuhrt** [comment]
I would love to see types and implementations separated. This is a trivial example that _hints_ at the problem within inline styles:

```js
const SpecifyResource = ({ create, destroy } : ResourceSpec) : Resource => (
```
versus
```js
 type SpecifyResource = ResourceSpec => Resource
const SpecifyResource = ({ create, destroy }) => (
```

My harsh opinion is that a more complicated signature and/or one utilizing more ES6 destructuring, spread, default assignment, editor line wrapping, etc. leads to kind of gestalt nonsense. Its just amazingly painful IMO. What keeps me using it anyways is how amazingly useful Flowtype can be overall of course! : )

**Comment by jasonkuhrt** [comment]
It also seems to me that allowing the user to rationalize over the types and implementations separately has value. I've heard to this referred to as _Type_ Driven Development. So for example in Haskell one might work out some function types leaving the implementations as `undefined` for as long as it makes sense to do so.

**Comment by samwgoldman** [comment]
Hi everyone. Thanks for thoughtfully discussing this issue among yourselves. I want to bring up my primary objection to the syntax proposed so far.

Currently, the statement `type a = b` creates a type alias, where the simple binding `a` can be substituted with the potentially more complex type expression `b`. For example, `type cb = () => {}` for the type of side-effecting callbacks.

After definition, the type alias can be used in a type position, potentially multiple times, to annotate values in the program. As @Gozala quoted me above, a type describes a universe of values.

The syntax proposed here co-opts this syntax and changes the meaning. If we implemented this feature using the proposed syntax, what would the `type a = b` statement mean? Can anyone provide a complete definition?

**Comment by jasonkuhrt** [comment]
@samwgoldman I'm not fully engaged at the conversation at hand so for instance the co-opting got me off guard. I can see what you are saying of course. My simple reply would be lets just not co-opt. From my limited experience in Haskell and my own intuition its the difference between `type alias` and `type of`. So maybe we could support:

```js
typeof SpecifyResource = ResourceSpec => Resource
 const SpecifyResource = ({ create, destroy }) => (
```

Or some such keyword; `typeof` might be a silly example.

**Comment by samwgoldman** [comment]
@jasonkuhrt Thanks for recommending an alternative.

I don't think `typeof` is a great choice because it is already an operator in JavaScript. The expression `typeof foo` already has meaning, so this proposal introduces a syntax ambiguity.

Let's also keep in mind that any proposal will also need to be implemented as a transform in Babel. This isn't a show stopper; just something to keep in mind.

**Comment by Gozala** [comment]
> The syntax proposed here co-opts this syntax and changes the meaning.

How does it changes this meaning ? 


> If we implemented this feature using the proposed syntax, what would the type a = b statement mean? Can anyone provide a complete definition?

I think it should remain the same although I am clearly missing something.


If I understand it correctly the issue is that currently types and implementations do share a same namespace:

/* @flow */

type sum = (x:number, y:number) => number
const sum = (x, y) => x = y

// sum name is already bound type sum
```

If they were not to share same namespace I believe the issues @samwgoldman is describing would not even exist. I also believe that is how it is in haskell or elm. That being said I don't know why the namespace is shared right & probably there is a reason.

**Comment by gyzerok** [comment]
@samwgoldman I am not expert but as far as I know
> If we implemented this feature using the proposed syntax, what would the type a = b statement mean?
This means that we define new type `a` which would be alias for `b`. The real difference is in case of
```js
type a = Number;
const a = 1
```
In the case when we do have a variable or function declaration on the next line with the same name we treat it of defined type.

**Comment by Gozala** [comment]
@gyzerok BTW in flow `Number` and `number` are different types see (http://flowtype.org/docs/quick-reference.html#the-number-constructor). Your examples seem to refer to `number`.

**Comment by gyzerok** [comment]
@Gozala yeah, you are right. I didnt use flow for a while now because of lack ES6 support.

**Comment by Gozala** [comment]
@gyzerok in fact with a @samwgoldman contributions most of ES6 is already supported by latest flow@0.17.0 ;)

**Comment by Gozala** [comment]
Ok so given the [response on IRC](https://botbot.me/freenode/flowtype/2015-10-15/?msg=51986938&page=3) I assume that type and implementation names do not necessarily have to share the same namespace, in which case I would like to propose following:

Keep type and implementation namespaces separate and in case of collisions require that implementation named `a` is of type `a`. By name collisions I mean they are defined in the same module or one is imported into other under same name or maybe both are imported into third module under the same name.


Further more current type annotated function:

```js
const sum = (x:number, y:number):number => x + y
```

Should be just a syntactic sugar for a following:

```js
type sum = (x:number, y:number) => number
const sum = (x, y) => x  +1
```

Note: If type annotated function is exported that should translate to both type and a implementation exports. In fact it is expected to be used multiple times in a case like described in earlier comment:
https://github.com/facebook/flow/issues/838#issuecomment-148191295


I believe that also does not change meaning of `type a = b`. Or to be specific it means / would mean exactly what @samwgoldman mentioned earlier:

>  the statement `type a = b` creates a type alias, where the simple binding `a` can be substituted with the potentially more complex type expression `b`.

Please note: That the type alias can be used in a type position, potentially multiple times, to annotate values in the program.

**Comment by Gozala** [comment]
Please note: I used named parameters in type definitions even though I thin that should be optional, but that making them optional is discussed in a separate thread #169

**Comment by unscriptable** [comment]
> It also seems to me that allowing the user to rationalize over the types and implementations separately has value. 

I agree.  On my very first time trying flow, I was looking for a way to separate the types from the implementation.  Types and implementations *should absolutely*  be rationalized separately for all but the simplest use cases.  This feature would help tremendously.

I eventually settled on using explicit assignment, similar to what @Gozala shows in one of his examples (and see my example below).  This seems to work, so I'm happily continuing down this path.  :)

A shopping cart example:

```js
type SKU = string;
type Item = { sku: SKU, quantity: number };

// cart states (implementation omitted)
class Empty {}
class Loaded {}
class Paid {}

type State = Empty | Loaded | Paid;
type Cart<State> = State;

// Note: type definition followed by implementation
type Create = () => Cart<Empty>;
export const create: Create = () => new Empty({});

// Again: type definition followed by implementation
type AddItem = (cart: Cart<Empty> | Cart<Loaded>, item: Item) => Cart<Loaded>;
export const addItem: AddItem = (cart, item) => {
    const items = cart.getItems();
    return new Loaded(items.addItem(item));
};
```

It feels a bit redundant to have to `create:Create` and `addItem:AddItem`, but this wouldn't be half bad if we could just *remove the parameter names from the function type definitions* (#169)!  

With respect to Haskell, yes there are separate namespaces for types and implementations.  It's probably best to think of them as two different sub-languages (similar to literal RegExp, ES6 templates, or JSX being separate, embedded languages in JavaScript).  The names in the type space are linked from within the implementation space via the function name.  The `::` symbol indicates you've crossed into the type sub-language+namespace and are defining the type information for the name or expression to the left.  Some simple examples are [here](https://github.com/nowaitapp/haskell-memcached/blob/master/Network/Memcached/Protocol.hs#L97-L111) if you're interested.

In short, Haskell ties a function's type to it's implementation with an explicit operator/symbol so there are no special rules to solve collisions or ambiguities.  Flow should be similarly explicit, imho, but would also love to see it be even more natural / less redundant.

**Comment by unscriptable** [comment]
Ok, I was wrong.  Flow seems to ignore the `type` line and infer it's own types from the implementation.  :disappointed: 

This should fail to type-check, too, iiuc, but it does not:

```js
export const wat: ((x:string) => number) = () => 5;
```

It seems like flow always overrides explicit type information with inferred implementation typing?  This seems backwards to me.

**Comment by samwgoldman** [comment]
@unscriptable In this case, the type `() => number` is a subtype of `(x:string) => number`. Because JavaScript semantics permit passing "extra" arguments to a function, any caller of the latter is also valid against the former.

**Comment by unscriptable** [comment]
For me, type information is a means to remove footguns from my code.  Footguns happen in a few different ways, the ones on my mind atm are: A) footguns that I, the code author, create by accident, B) footguns that *code users* create by [mis]using my code, and C) the footguns that "downstream" *code authors* create by extending my code.  It seems that by allowing implementation to implicitly override explicit type info, flow minimizes B, but not A or C (unless there's some way to prevent the implicit inferred type overrides).

I'm a bit dismayed by the implicit overrides, because I (a code author) need to work really hard (and likely would have to continue to use something like `flow suggest` constantly) to understand if the type system is doing what I actually want it to do.

Again, this feels backwards.  To my mind, my types should enforce that my code is correct.  If I can prove that my types are correct, then my code will be correct, too (limited to the amount of "correctness" the type system offers).  

Separation of types from implementation (as suggested by @Gozala and the main topic of this github issue) is helpful, if not critical, for me to reason about my types so I can prove that they are correct.  

Thoughts?

**Comment by samwgoldman** [comment]
I'm sympathetic about the extra arguments behavior, as it can be unexpected, but I think your argument is based on a false premise. You said that flow "allows implementation to implicitly override explicit type info," but that just isn't the case.

Given the expression `(y: T)`, Flow enforces that the expression `y` is compatible with type `T`. This behavior is also found in variable declarations (`var x: T = y`) and other scenarios where expressions flow into a typed context (passing arguments to typed parameters, say).

Can you give an example of what you consider an "implicit override" so I might understand what you mean?

**Comment by samwgoldman** [comment]
OK, so this issue has been open for a long time. I've read everyone's comments here, and I really appreciate everyone taking the time to express their feelings on this issue.

Ultimately, I'm going to close this. We have no plans to change Flow to support the kind of syntax proposed here.

If the proposed syntax changes/additions increased the expressiveness of our type system, I think there would be more to say here, but from what I can see it's a matter of taste. What we have now looks different, but provides 100% of the same functionality.

There's no accounting for taste and we aren't going to please everyone with syntax. What we have now works and, importantly, converges nicely with TypeScript's syntax.

Thanks again for everyone's thoughtful comments.

**Comment by jasonkuhrt** [comment]
@samwgoldman Thanks for being frank. I disagree but alas so be it : )

**Comment by Gozala** [comment]
@samwgoldman I think your completely ignoring the fact that there are cases which can only be typed separate from implementation see #1318 for specific example. Although I guess variadic functions may also not be something that flow would like to support.
